bin/katello-certs-check.rb

```diff
@@ -0,0 +1,175 @@
+  end
+  result
+end
+
+cert_path = nil
+key_path = nil
+bundle_path = nil
+
+parser = OptionParser.new do |opts|
+  opts.on('-t [SERVER|PROXY]', String, 'Target') do |v|
+    # TODO store?
+  end
+
+  opts.on('-c [CERTIFICATE]', String, 'Path to certificate file') do |path|
+    unless File.exist?(path)
+      fail "Path '#{path}' does not exist"
+    end
+    cert_path = path
+  end
+
+  opts.on('-k [KEY]', String, 'Path to private key file') do |path|
+    unless File.exist?(path)
+      fail "Path '#{path}' does not exist"
+    end
+    key_path = path
+  end
+
+  opts.on('-b [BUNDLE]', String, 'Path to Certificate Authority bundle file') do |path|
+    unless File.exist?(path)
+      fail "Path '#{path}' does not exist"
+    end
+    bundle_path = path
+  end
+end
+
+parser.parse!
+
+if [cert_path, key_path, bundle_path].include?(nil)
+  # TODO: print usage
+  fail "Missing cert, key or bundle option"
+end
+
+cert = check("Reading certificate file '#{cert_path}'") do
+  OpenSSL::X509::Certificate.new(File.read(cert_path))
+rescue Exception => e
+  raise FailedCheck, "Failed to read certificate: #{e}"
+  # TODO: check if file is DER encoded
+end
+
+key = check("Reading private key '#{key_path}'") do
+  OpenSSL::PKey.read(File.read(key_path), '')
+rescue Exception => e
+  if e.message.include?('bad decrypt')
+    raise FailedCheck, "Unable to parse private key '#{key_path}'. Likely contains a passphrase"
+  end
+  raise FailedCheck, "Failed to read private key: #{e}"
+end
+
+bundle = check("Reading CA certificate '#{bundle_path}'") do
+  OpenSSL::X509::Certificate.new(File.read(bundle_path))
+rescue Exception => e
+  raise FailedCheck, "Failed to read CA certificate: #{e}"
+end
+
+check('Checking certificate expiration') do
+  now = Time.now
+  if cert.not_before > now
+    raise FailedCheck, "Certificate not valid yet: #{cert.not_before} is in the future"
+  end
+
+  if cert.not_after < now
+    raise FailedCheck, "Certificate no longer valid: #{cert.not_after} is in the past"
+  end
+end
+
+check('Checking CA bundle expiration') do
+  now = Time.now
+  if bundle.not_before > now
+    raise FailedCheck, "Certificate not valid yet: #{bundle.not_before} is in the future"
+  end
+
+  if bundle.not_after < now
+    raise FailedCheck, "Certificate no longer valid: #{bundle.not_after} is in the past"
+  end
+end
+
+check('Verifying if private key matches certificate') do
+  unless cert.check_private_key(key)
+    raise FailedCheck, "Certificate '#{cert_path}' is not signed by '#{key_path}"
+  end
+end
+
+store = OpenSSL::X509::Store.new
+# TODO: also for clients?
+store.purpose = OpenSSL::X509::PURPOSE_SSL_SERVER
+store.add_file(bundle_path)
+
+chain = check('Verify certificate is singed by bundle') do
+  unless store.verify(cert)
+    raise FailedCheck, "Certificate is not signed by bundle: #{store.error_string}"
+  end
+  store.chain
+end
+
+# TODO
+check('Verifying root is included in bundle has CA:TRUE') do
+  # TODO: oid for basicConstraints?
```

Comment:
I get the impression you can rely on the name, but I wonder if that's always true.